openfe_gromacs/protocols/gromacs_md/md_methods.py

```diff
@@ -0,0 +1,649 @@
+            protocol_settings.solvation_settings
+        )
+        charge_settings: BasePartialChargeSettings = (
+            protocol_settings.partial_charge_settings
+        )
+        sim_settings_em: EMSimulationSettings = protocol_settings.simulation_settings_em
+        sim_settings_nvt: NVTSimulationSettings = (
+            protocol_settings.simulation_settings_nvt
+        )
+        sim_settings_npt: NPTSimulationSettings = (
+            protocol_settings.simulation_settings_npt
+        )
+        output_settings_em: EMOutputSettings = protocol_settings.output_settings_em
+        output_settings_nvt: NVTOutputSettings = protocol_settings.output_settings_nvt
+        output_settings_npt: NPTOutputSettings = protocol_settings.output_settings_npt
+        integrator_settings = protocol_settings.integrator_settings
+
+        solvent_comp, protein_comp, small_mols = system_validation.get_components(
+            stateA
+        )
+
+        # Raise an error when no SolventComponent is provided as this Protocol
+        # currently does not support vacuum simulations
+        if not solvent_comp:
+            errmsg = (
+                "No SolventComponent provided. This protocol currently does"
+                "not support vacuum simulations."
+            )
+            raise ValueError(errmsg)
+
+        # 1. Write out .mdp files
+        mdps = []
+        if protocol_settings.simulation_settings_em.nsteps > 0:
+            settings_dict = (
+                sim_settings_em.dict()
+                | output_settings_em.dict()
+                | pre_defined_settings
+                | pre_defined_settings_em
+            )
+            mdp = _dict2mdp(settings_dict, shared_basepath)
+            mdps.append(mdp)
+        if protocol_settings.simulation_settings_nvt.nsteps > 0:
+            settings_dict = (
+                sim_settings_nvt.dict()
+                | output_settings_nvt.dict()
+                | pre_defined_settings
+            )
+            mdp = _dict2mdp(settings_dict, shared_basepath)
+            mdps.append(mdp)
+        if protocol_settings.simulation_settings_npt.nsteps > 0:
+            settings_dict = (
+                sim_settings_npt.dict()
+                | output_settings_npt.dict()
+                | pre_defined_settings
+            )
+            mdp = _dict2mdp(settings_dict, shared_basepath)
+            mdps.append(mdp)
+
+        # 2. Create stateA system
+        # Create a dictionary of OFFMol for each SMC for bookkeeping
+        smc_components: dict[SmallMoleculeComponent, OFFMolecule]
+
+        smc_components = {i: i.to_openff() for i in small_mols}
+
+        # a. assign partial charges to smcs
+        self._assign_partial_charges(charge_settings, smc_components)
+
+        # b. Validate that no constraints are specified which is needed
+        #    for Interchange to work properly
+        if forcefield_settings.constraints:
```

Comment:
We should validate this before the Unit, either at Protocol creation or we should subclass the settings and add a validator?

Reply:
I made a subclass of the `OpenMMSystemGeneratorFFSettings` and added a validator.